client/peerset/src/lib.rs

```diff
@@ -235,61 +240,92 @@ impl Peerset {
 	}
 
 	fn on_add_reserved_peer(&mut self, peer_id: PeerId) {
-		let mut reserved = self.data.get_priority_group(RESERVED_NODES).unwrap_or_default();
-		reserved.insert(peer_id);
-		self.data.set_priority_group(RESERVED_NODES, reserved);
-		self.alloc_slots();
+		self.on_add_to_priority_group(RESERVED_NODES, peer_id);
 	}
 
 	fn on_remove_reserved_peer(&mut self, peer_id: PeerId) {
-		let mut reserved = self.data.get_priority_group(RESERVED_NODES).unwrap_or_default();
-		reserved.remove(&peer_id);
-		self.data.set_priority_group(RESERVED_NODES, reserved);
-		match self.data.peer(&peer_id) {
-			peersstate::Peer::Connected(peer) => {
-				if self.reserved_only {
-					peer.disconnect();
-					self.message_queue.push_back(Message::Drop(peer_id));
-				}
-			}
-			peersstate::Peer::NotConnected(_) => {},
-			peersstate::Peer::Unknown(_) => {},
-		}
+		self.on_remove_from_priority_group(RESERVED_NODES, peer_id);
 	}
 
 	fn on_set_reserved_only(&mut self, reserved_only: bool) {
 		self.reserved_only = reserved_only;
-		self.data.set_priority_only(reserved_only);
 
 		if self.reserved_only {
-			// Disconnect non-reserved nodes.
-			let reserved = self.data.get_priority_group(RESERVED_NODES).unwrap_or_default();
+			// Disconnect all the nodes that aren't reserved.
 			for peer_id in self.data.connected_peers().cloned().collect::<Vec<_>>().into_iter() {
+				if self.priority_groups.get(RESERVED_NODES).into_iter().flatten().any(|p| *p == peer_id) {
+					continue;
+				}
+
 				let peer = self.data.peer(&peer_id).into_connected()
 					.expect("We are enumerating connected peers, therefore the peer is connected; qed");
-				if !reserved.contains(&peer_id) {
-					peer.disconnect();
-					self.message_queue.push_back(Message::Drop(peer_id));
-				}
+				peer.disconnect();
+				self.message_queue.push_back(Message::Drop(peer_id));
 			}
+
 		} else {
 			self.alloc_slots();
 		}
 	}
 
 	fn on_set_priority_group(&mut self, group_id: &str, peers: HashSet<PeerId>) {
-		self.data.set_priority_group(group_id, peers);
-		self.alloc_slots();
+		// Determine the difference between the current group and the new list.
+		let (to_insert, to_remove) = {
+			let current_group = self.priority_groups.entry(group_id.to_owned()).or_default();
+			let to_insert = peers.iter().filter(|peer| !current_group.iter().any(|p| p == *peer))
+				.cloned().collect::<Vec<_>>();
+			let to_remove = current_group.iter().filter(|peer| !peers.iter().any(|p| p == *peer))
```

Comment:
```suggestion
			let to_remove = current_group.iter().filter(|peer| !peers.contains(*peer))
```
`HashSet::difference` could also work here